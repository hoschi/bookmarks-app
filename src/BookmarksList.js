import * as R from 'ramda'
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import List from '@material-ui/core/List'
import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'
import Divider from '@material-ui/core/Divider'
import Paper from '@material-ui/core/Paper'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'

const styles = R.always({
    link: {
        color: 'inherit',
        textDecoration: 'none',
    },
    linkRead: {
        color: '#00000080',
    },
})

function BookmarksList({ classes }) {
    return (
        <Query
            query={gql`
                {
                    allBookmarks(sortField: "isRead", sortOrder: "ASC") {
                        id
                        title
                        url
                        isRead
                    }
                }
            `}
        >
            {({ loading, error, data }) => {
                if (loading) return <p>Loading...</p>
                if (error) return <p>Error {error.toString()}</p>
                const { allBookmarks } = data

                return (
                    <Paper>
                        <Typography component="div">
                            <List disablePadding>
                                {allBookmarks.map(({ id, title, url, isRead }, i) => (
                                    <Fragment key={id}>
                                        <ListItem component="a" href={url} className={classes.link}>
                                            <ListItemText
                                                primary={title}
                                                classes={{
                                                    primary: classNames({
                                                        [classes.linkRead]: isRead,
                                                    }),
                                                }}
                                            />
                                        </ListItem>
                                        {i !== allBookmarks.length - 1 && <Divider />}
                                    </Fragment>
                                ))}
                            </List>
                        </Typography>
                    </Paper>
                )
            }}
        </Query>
    )
}

BookmarksList.propTypes = {
    classes: PropTypes.object.isRequired,
}

export default withStyles(styles)(BookmarksList)
