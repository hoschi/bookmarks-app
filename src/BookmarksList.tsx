import * as React from 'react'
import { Typography, List, Divider, Paper } from '@mui/material'
import { queryAllBookmarks } from './gql'
import { useQuery } from '@apollo/client'
import { BookmarkItem } from './BookmarkItem'

export const BookmarksList: React.FC = () => {
    const { loading, error, data } = useQuery(queryAllBookmarks)
    if (loading) return <p>Loading...</p>
    if (error) return <p>Query Error {error.toString()}</p>

    const bookmarks = data?.allBookmarks || []
    return (
        <Paper>
            <Typography component="div">
                <List disablePadding>
                    {bookmarks.map((item, i: number) => {
                        // why can this be null? I gues this is, because json-graphql-server returns [Bookmark] and not [Bookmark!]. See https://www.apollographql.com/blog/using-nullability-in-graphql#nullability-and-lists
                        if (!item) {
                            throw new Error(`items from server shouldn't be null`)
                        }

                        return (
                            <React.Fragment key={item.id}>
                                <BookmarkItem bookmark={item} />
                                {i !== bookmarks.length - 1 && <Divider />}
                            </React.Fragment>
                        )
                    })}
                </List>
            </Typography>
        </Paper>
    )
}
