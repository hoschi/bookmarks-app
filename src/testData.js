const addHours = require('date-fns/add_hours')
let currentDate = new Date('2017-01-02T12:00:00')

function getId() {
    const r = currentDate
    currentDate = addHours(currentDate, 1)
    return r.getTime().toString()
}

export const bookmarksList = [
    {
        id: getId(),
        title: '🎼webpack 4: released today!!✨',
        url: 'https://medium.com/webpack/webpack-4-released-today-6cdb994702d4',
        isRead: false,
    },
    {
        id: getId(),
        title: 'mg979/vim-visual-multi: Multiple cursors project for vim/neovim (wip)',
        url: 'https://github.com/mg979/vim-visual-multi',
        isRead: false,
    },
    {
        id: getId(),
        title: 'How To GraphQL - Mutations',
        url: 'https://www.howtographql.com/react-apollo/3-mutations-creating-links/',
        isRead: true,
    },
    {
        id: getId(),
        title: 'Check out new Neovim UI build with JavaScript - Veonim',
        url: 'https://github.com/veonim/veonim',
        isRead: false,
    },
    {
        id: getId(),
        title: '🍾🚀 webpack 3: Official Release!! 🚀🍾',
        url: 'https://medium.com/webpack/webpack-3-official-release-15fd2dd8f07b',
        isRead: true,
    },
]
