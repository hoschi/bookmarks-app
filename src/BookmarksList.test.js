import React from 'react'
import { shallow } from 'enzyme'
import { Query } from 'react-apollo'
import { bookmarksList } from './testData'
import BookmarksList from './BookmarksList'

test('render', function() {
    const actual = shallow(<BookmarksList />)
    expect(actual).toMatchSnapshot()
})

describe('Query queryAllBookmarks', function() {
    const actual = shallow(<BookmarksList />)
        .find(Query)
        .prop('children')

    it('loading', function() {
        const queried = shallow(actual({ loading: true }))
        expect(queried).toMatchSnapshot()
    })

    it('error', function() {
        const queried = shallow(actual({ error: new Error('uh oh') }))
        expect(queried).toMatchSnapshot()
    })

    it('with data', function() {
        const queried = shallow(actual({ data: { allBookmarks: bookmarksList } }))
        expect(queried).toMatchSnapshot()
    })
})
