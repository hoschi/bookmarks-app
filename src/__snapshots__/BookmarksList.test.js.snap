// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query queryAllBookmarks error 1`] = `
<p>
  Query Error 
  Error: uh oh
</p>
`;

exports[`Query queryAllBookmarks loading 1`] = `
<p>
  Loading...
</p>
`;

exports[`Query queryAllBookmarks with data 1`] = `
<Paper
  classes={
    Object {
      "elevation0": "MuiPaper-elevation0-3",
      "elevation1": "MuiPaper-elevation1-4",
      "elevation10": "MuiPaper-elevation10-13",
      "elevation11": "MuiPaper-elevation11-14",
      "elevation12": "MuiPaper-elevation12-15",
      "elevation13": "MuiPaper-elevation13-16",
      "elevation14": "MuiPaper-elevation14-17",
      "elevation15": "MuiPaper-elevation15-18",
      "elevation16": "MuiPaper-elevation16-19",
      "elevation17": "MuiPaper-elevation17-20",
      "elevation18": "MuiPaper-elevation18-21",
      "elevation19": "MuiPaper-elevation19-22",
      "elevation2": "MuiPaper-elevation2-5",
      "elevation20": "MuiPaper-elevation20-23",
      "elevation21": "MuiPaper-elevation21-24",
      "elevation22": "MuiPaper-elevation22-25",
      "elevation23": "MuiPaper-elevation23-26",
      "elevation24": "MuiPaper-elevation24-27",
      "elevation3": "MuiPaper-elevation3-6",
      "elevation4": "MuiPaper-elevation4-7",
      "elevation5": "MuiPaper-elevation5-8",
      "elevation6": "MuiPaper-elevation6-9",
      "elevation7": "MuiPaper-elevation7-10",
      "elevation8": "MuiPaper-elevation8-11",
      "elevation9": "MuiPaper-elevation9-12",
      "root": "MuiPaper-root-1",
      "rounded": "MuiPaper-rounded-2",
    }
  }
  component="div"
  elevation={2}
  square={false}
>
  <WithStyles(Typography)
    component="div"
  >
    <WithStyles(List)
      disablePadding={true}
    >
      <fromRenderProps(WithStyles(BookmarkItem))
        id="1483354800000"
        isRead={false}
        title="🎼webpack 4: released today!!✨"
        url="https://medium.com/webpack/webpack-4-released-today-6cdb994702d4"
      />
      <WithStyles(Divider) />
      <fromRenderProps(WithStyles(BookmarkItem))
        id="1483358400000"
        isRead={false}
        title="mg979/vim-visual-multi: Multiple cursors project for vim/neovim (wip)"
        url="https://github.com/mg979/vim-visual-multi"
      />
      <WithStyles(Divider) />
      <fromRenderProps(WithStyles(BookmarkItem))
        id="1483362000000"
        isRead={true}
        title="How To GraphQL - Mutations"
        url="https://www.howtographql.com/react-apollo/3-mutations-creating-links/"
      />
      <WithStyles(Divider) />
      <fromRenderProps(WithStyles(BookmarkItem))
        id="1483365600000"
        isRead={false}
        title="Check out new Neovim UI build with JavaScript - Veonim"
        url="https://github.com/veonim/veonim"
      />
      <WithStyles(Divider) />
      <fromRenderProps(WithStyles(BookmarkItem))
        id="1483369200000"
        isRead={true}
        title="🍾🚀 webpack 3: Official Release!! 🚀🍾"
        url="https://medium.com/webpack/webpack-3-official-release-15fd2dd8f07b"
      />
    </WithStyles(List)>
  </WithStyles(Typography)>
</Paper>
`;

exports[`render 1`] = `
<Query
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "sortField",
                    },
                    "value": Object {
                      "block": false,
                      "kind": "StringValue",
                      "value": "isRead",
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "sortOrder",
                    },
                    "value": Object {
                      "block": false,
                      "kind": "StringValue",
                      "value": "ASC",
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "allBookmarks",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                    },
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "title",
                      },
                    },
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "url",
                      },
                    },
                    Object {
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "isRead",
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 153,
        "source": Object {
          "body": "
    {
        allBookmarks(sortField: \\"isRead\\", sortOrder: \\"ASC\\") {
            id
            title
            url
            isRead
        }
    }
",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "GraphQL request",
        },
        "start": 0,
      },
    }
  }
>
  <Component />
</Query>
`;
